Refresh.GameServer/Importing/AssetImporter.cs

```diff
@@ -19,80 +21,81 @@ public AssetImporter(Logger? logger = null, IDateTimeProvider? timeProvider = nu
-            if (path.Contains('/') && !isPsp) continue;
+            int threadId = i;
+            new Thread(() =>
+            {
+                GameDatabaseContext database = databaseProvider.GetContext();
 
-            string hash = isPsp ? path[4..] : path;
-            
-            byte[] data = dataStore.GetDataFromStore(path);
-            
-            GameAsset? newAsset = this.ReadAndVerifyAsset(hash, data, isPsp ? TokenPlatform.PSP : null, database);
-            if (newAsset == null) continue;
+                List<GameAsset> assets = new(assetHashes.Length / threadCount);
 
-            GameAsset? oldAsset = database.GetAssetFromHash(hash);
+                while (hashesToProcess.TryTake(out string? path))
+                {
+                    bool isPsp = path.StartsWith("psp/");
+                    //If the hash has a `/` and it doesnt start with `psp/`, then its an invalid asset
+                    if (path.Contains('/') && !isPsp) continue;
 
-            if (oldAsset != null)
-            {
-                newAsset.OriginalUploader = oldAsset.OriginalUploader;
-                newAsset.UploadDate = oldAsset.UploadDate;
-                updatedAssets++;
-            }
-            else
-            {
-                newAssets++;
-            }
+                    string hash = isPsp ? path[4..] : path;
+            
+                    byte[] data = dataStore.GetDataFromStore(path);
+            
+                    GameAsset? newAsset = this.ReadAndVerifyAsset(hash, data, isPsp ? TokenPlatform.PSP : null, database);
+                    if (newAsset == null) continue;
+
+                    GameAsset? oldAsset = database.GetAssetFromHash(hash);
+
+                    if (oldAsset != null)
+                    {
+                        newAsset.OriginalUploader = oldAsset.OriginalUploader;
+                        newAsset.UploadDate = oldAsset.UploadDate;
+                        Interlocked.Increment(ref updatedAssets);
+                    }
+                    else
+                    {
+                        Interlocked.Increment(ref newAssets);
+                    }
+                    
+                    assets.Add(newAsset);
+                    this.Info($"[{Interlocked.Increment(ref processed)}/{assetHashes.Length}] Processed {newAsset.AssetType} asset {hash} ({AssetSafetyLevelExtensions.FromAssetType(newAsset.AssetType, newAsset.AssetFormat)})");
+                }
+                
+                database.AddOrUpdateAssetsInDatabase(assets);
 
-            assets.Add(newAsset);
-            this.Info($"Processed {newAsset.AssetType} asset {hash} ({AssetSafetyLevelExtensions.FromAssetType(newAsset.AssetType, newAsset.AssetFormat)})");
+                tasksDone[threadId] = true;
+            }).Start();
         }
-        
-        database.AddOrUpdateAssetsInDatabase(assets);
 
+        SpinWait.SpinUntil(() => tasksDone.All(done => done));
```

Comment:
Wouldn't this hang if a thread throws internally and exits? Also disagree with spinning on a long-running operation.

It's better if we use tasks here. Store `Task[]` instead of `bool[]`, then use [`Task.WaitAll`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitall?view=net-8.0#system-threading-tasks-task-waitall(system-threading-tasks-task())), which also has the bonus of handling `Task`s that throw by bubbling up an `AggregateException`. It's true that we don't disturb the thread pool by using dedicated threads, but we can also just hint to the pool that the task will be long-running and it'll be handled much better.